src/hotspot/share/runtime/continuationFreezeThaw.cpp

```diff
@@ -565,10 +565,49 @@ void FreezeBase::copy_to_chunk(intptr_t* from, intptr_t* to, int size) {
 #endif
 }
 
+static void assert_frames_in_continuation_are_safe(JavaThread* thread) {
+#ifdef ASSERT
+  ContinuationEntry* ce = thread->last_continuation();
+  RegisterMap map(thread,
+                  RegisterMap::UpdateMap::include,
+                  RegisterMap::ProcessFrames::skip,
+                  RegisterMap::WalkContinuation::skip);
+  map.set_include_argument_oops(false);
+  for (frame f = thread->last_frame(); Continuation::is_frame_in_continuation(ce, f); f = f.sender(&map)) {
+    StackWatermark* watermark = StackWatermarkSet::get(thread, StackWatermarkKind::gc);
+    if (watermark != nullptr) {
+      watermark->assert_is_frame_safe(f);
+    }
+  }
+#endif // ASSERT
+}
+
+#ifdef ASSERT
+static bool monitors_on_stack(JavaThread* thread) {
+  assert_frames_in_continuation_are_safe(thread);
+  ContinuationEntry* ce = thread->last_continuation();
+  RegisterMap map(thread,
+                  RegisterMap::UpdateMap::include,
+                  RegisterMap::ProcessFrames::skip,
+                  RegisterMap::WalkContinuation::skip);
+  map.set_include_argument_oops(false);
+  for (frame f = thread->last_frame(); Continuation::is_frame_in_continuation(ce, f); f = f.sender(&map)) {
```

Comment:
We walk the stack twice - once to assert_is_frame_safe() for the stack watermark and then this set of calls about owning locks.  Why not just add if (watermark != nullptr) watermark->assert_is_frame_safe(f); here.  Why is the same thing twice - ie getting the RegisterMap?
I thought the problem with monitors_on_stack was that the watermark was not safe for older (higher) frames?  Is it that at this call location, the stack watermark has been processed?

Reply:
Am I that predictable.  Okay, I see the double code/loop is temporary and one has future usefulness.